### PR TITLE
docs: outline the release cycle in depth

### PR DESCRIPTION
**Pull request by brianp**
Description
---
Outline in more detail how version tags are managed and the process of releasing a new tag for a particular channel and network.

Motivation and Context
---
Make sure everyones on the same page regarding tagging best practices, branching, and releasing.

How Has This Been Tested?
---
It will likely be read, and discussed.

### CODE REVIEW

File: docs/src/branching_releases.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+- TestNet
+- NextNet
+- StageNet/MainNet
+
+### Networks
+
+- MainNet: The big shebang, the head hancho, the real deal. This is the Tari protocol at work.
+- StageNet: A stable network as similar as possible to MainNet available for testing.
+- NextNet: The next realease of the network. This should become available 8 weeks prior to going live on MainNet.
```
Comment (CjS77) [inline docs/src/branching_releases.md]:
```suggestion
- NextNet: The next release of the network. This should become available 8 weeks prior to going live on MainNet.
```

File: docs/src/branching_releases.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+## Releases
+
+Let's define some facts for our discussion about releasing.
+
+- StageNet/MainNet is at v1.53.0
+- NextNet is at v1.54.0-rc.0
+- TestNet is at v1.55.0-pre.4
+
+The naming practice here helps to differentiate where any particular version is meant to be deployed too. Any version containing `pre` is destined for a TestNet and will have all features enabled. A version containing `rc` is destined for NextNet as it is the release candidate for the future StageNet/MainNet release. Any singular version with no pre-release version appended such as v1.53.0 means this is a StageNet/MainNet build. No in development features should be enabled when this is compiled.
```
Comment (CjS77) [inline docs/src/branching_releases.md]:
```suggestion
The naming practice here helps to differentiate where any particular version is meant to be deployed too. Any version containing `pre` is destined for a TestNet and will have all features enabled. A version containing `rc` is destined for NextNet as it is the release candidate for the future StageNet/MainNet release. Any singular version with no pre-release version appended such as v1.53.0 means this is a StageNet/MainNet build. None of the in-development features should be enabled when this is compiled.
```

File: docs/src/branching_releases.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+### Development releases
+
+A developer writes a new feature `Feature-Y`. It is written behind a feature gate. They merge this new feature into `development`. They now want to run this feature on the `Esme` test network. They create a disposable release branch in the name of the version that will be tagged with the preface `testnet` and incrementing the number at the end of the `pre` version `testnet-1.55-pre.5`. Once any upgrade related changes are made such as incrementing the version through different crates, the developer will then tag the version `v1.55.0-pre.5`. The branch can be merged back into development, and the branch deleted.
```
Comment (CjS77) [inline docs/src/branching_releases.md]: This paragraph isn't terribly clear.

esp this clause, "a disposable release branch in the name of the version that will be tagged with the preface `testnet`"

Could you try rewriting the paragraph?

**Comment by brianp** [comment]
This needs more disambiguation for _what version numbers mean_.

MAJOR being for hardforks
MINOR for breaking changes
PATCH for all other updates _including_ hotfixes.

--- 

There's a notable difference here from a pattern we use now where we do not need to manage specific branches for master, stagenet, nextnet etc. Tags in the correct places should be sufficient, long lived branches for these channels don't serve much purpose in this process. 

If we want to follow gitflow _closer_ then we can reintroduce the maintenance of these branches by making every release branch merge into their channel specific branch. It's still worth noting that things only ever go into those branches, and hotfixes from them. No other work would be done on them. But also that's only if we actually wanted to maintain them.

File: docs/src/branching_releases.md (hunk 3 above)
Comment (brianp) [reply docs/src/branching_releases.md]: Yup will do. It needs some work.

File: docs/src/branching_releases.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+We run a series of networks for different purposes but each network will have come from one of these three channels.
+
+- TestNet
+- NextNet
+- StageNet/MainNet
+
+### Networks
+
+- MainNet: The big shebang, the head hancho, the real deal. This is the Tari protocol at work.
```
Comment (SWvheerden) [inline docs/src/branching_releases.md]:
```suggestion
- MainNet: The big shebang, the head honcho, the real deal. This is the Tari protocol at work.
```

File: docs/src/branching_releases.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+- StageNet: A stable network as similar as possible to MainNet available for testing.
+- NextNet: The next release of the network. This should become available 8 weeks prior to going live on MainNet.
+- TestNet(s): Our named test networks used for development. Where we can play a little fast and loose with features where, although we try to avoid it- the networks may see breakages and resets more frequently.
+  - Example networks (non exhaustive):
```
Comment (SWvheerden) [inline docs/src/branching_releases.md]:
```suggestion
  - Example networks (non-exhaustive):
```

File: docs/src/branching_releases.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+Next the developer will checkout the development branch. From here create a new release branch `nextnet-1.55.0-rc.0`. Make any version bumps or changes necessary. Tag this new branch as v1.55.0-rc.0 triggering a CI compilation of the NextNet binary with the flag `TARI_NETWORK=nextnet` which will allow feature gates in the testing status to stay open. The release branch can be merged back into development and then deleted.
+Lastly the development branch should be reset to the _next_ version v1.56.0-pre.0. This make it simple for the next testnet release and leaves no questions as to what version comes next.
```
Comment (SWvheerden) [inline docs/src/branching_releases.md]:
```suggestion
Lastly the development branch should be reset to the _next_ version v1.56.0-pre.0. This makes it simple for the next testnet release and leaves no questions as to what version comes next.
```

**Review by SWvheerden — DISMISSED** [review]
ACK
Looks good I agree with the process

**Comment by brianp** [comment]
@stringhandler ready to go

**Review by stringhandler — APPROVED** [review]

**Pull request merged**